### DISCUSSION TITLE
Bad performance on large dataset with relationships (manual query is much better!)

### DISCUSSION THREAD

**Discussion by leon-wbr** [directus/directus#15745] [Archive]
I have a large dataset (25k rows) with a specific one-to-many relationship that I need to fetch with it. This manual query runs in 150-200ms:

```sql
SELECT 
	places.id, 
	places.position,
	ROUND(AVG(Ratings.value), 1) AS rating
FROM places LEFT JOIN ratings
	ON places.id = ratings.place
GROUP BY places.id;
```

However, doing the same request via Directus with just the request `/items/places?fields=id,position,ratings.value,updated_at&limit=-1` ends up taking more than 20 seconds. Why is this?

Is there a way to use my manual query instead? Or an even more efficient way to achieve what I need?

**Comment by azrikahar** [comment]
> However, doing the same request via Directus with just the request /items/places?fields=id,position,ratings.value,updated_at&limit=-1 ends up taking more than 20 seconds. Why is this?

I'm not sure why this may happen, but would you mind helping to diagnose this by changing the `LOG_LEVEL` environment variable to `trace`? You should then be able to see all SQL queries performed directly shown in the logs, and can you pinpoint and share the SQL query generated for this request? It'd be very helpful to find out what could be the potential cause, thanks in advance!

> Is there a way to use my manual query instead? Or an even more efficient way to achieve what I need?

You should be able to create a [custom endpoint extension](https://docs.directus.io/extensions/endpoints.html#register-function) and use `database` in the context which is a [Knex](https://knexjs.org/) instance and form your own query.

**Reply by leon-wbr** [reply]
These are the queries I can see in the console. I'm not sure why ratings are fetched multiple times? You can see the query at the bottom. I've removed the huge amount of IDs that are put in there...

```sql
23:04:54 ≡ƒöì [41.790ms] select `places`.`id` from `places` order by `places`.`id` asc []
23:04:54 ≡ƒöì [101.453ms] select `ratings`.`value`, `ratings`.`place`, `ratings`.`id` from `ratings` where `ratings`.`place` in (?, ..., ?) order by `ratings`.`id` asc limit ? [1, 2, 3, 4...25515, 25516, 25517, 25000]
23:05:23 ≡ƒöì [115.529ms] select `ratings`.`value`, `ratings`.`place`, `ratings`.`id` from `ratings` where `ratings`.`place` in (?, ..., ?) order by `ratings`.`id` asc limit ? offset ? [1, 2, 3, 4...25517, 25000, 25000]
23:05:51 ≡ƒöì [52.047ms] select `ratings`.`value`, `ratings`.`place`, `ratings`.`id` from `ratings` where `ratings`.`place` in (?, ..., ?) order by `ratings`.`id` asc limit ? offset ? [1, 2, 3, 4...25517, 25000, 50000]
23:05:54 Γ£¿ request completed GET 200 /items/places?fields=id,ratings.value&limit=-1 1m 0.2s
```

I did use a custom endpoint but this still seems like odd behavior. My custom endpoint gets this:

```sql
23:12:52 ≡ƒöì [145.662ms]
                        SELECT
                                places.id,
                                places.position,
                                ROUND(AVG(ratings.value), 1) as rating,
                                places.updated_at
                        FROM places LEFT JOIN ratings
                                ON places.id = ratings.place
                        GROUP BY places.id;
                 []
23:12:52 Γ£¿ request completed GET 200 /places 184ms
```

The time it takes to execute the query is quite similar, but you can see that the request above takes a total of 1m to complete whereas the manual query only takes about 200ms.

If there is anything else I can provide you with, please tell me. Might take a little bit for me to get back to you but I'll do my best.

**Comment by br41nslug** [comment]
It takes 20 seconds for that response? seems like there might be more going on 🤔 Just to cover the bases: Do you have any flows or extensions that could be impacting performance here? like the `items.query` or `items.read` events for instance

**Reply by leon-wbr** [reply]
I have posted my console output in the comment above and it logs the response taking about a minute. I didn't implement any custom extensions that should impact this, what I have are actions on create and update. Those should not fire and don't seem to be firing. I removed them and the result is the same.
